Express noncopyability through the suppression of `Copyable`

kavon opened the pull request:
See https://forums.swift.org/t/second-review-se-0390-noncopyable-structs-and-enums/63866

This implements `~Copyable` constraints, which means "suppress the implicit conformance to `Copyable` for this type"

That's not the final accepted syntax; writing `without Copyable` appears to have the most support on the forums. A follow-up PR will change this syntax once the dust has settled from the review.

companion PR: https://github.com/apple/swift-syntax/pull/1588

resolves rdar://106775103

kavon committed:
add missing undef in metaprogramming

kavon committed:
add a notion of suppressible known protocols

kavon committed:
rename `_Copyable` to `Copyable` and make it available

kavon committed:
implement `~Copyable` to suppress implicit conformance to `Copyable`

Also has been referred to as "without Copyable"

See SE-390 for details, or read the discussion here:

https://forums.swift.org/t/second-review-se-0390-noncopyable-structs-and-enums/63866

kavon committed:
[tests] use `~Copyable` instead of `@_moveOnly`

many of these are mechanical changes

kavon commented:
@swift-ci please smoke test

kavon commented:
@swift-ci please smoke test

kavon commented:
@swift-ci smoke test

kavon committed:
[tests] improve a test to cover ~Copyable

kavon commented:
@swift-ci smoke test macOS

kavon commented:
@swift-ci please clean smoke test

kavon committed:
[tests] improve the parse-stdlib test

kavon committed:
[tests] add very basic test to ensure modules / interfaces for moveonly types work with library evolution

kavon committed:
[tests] remove outdated FIXME

kavon committed:
prevent uses of `Copyable`; fix bogus errors when `~Copyable` appears

Rather than try to omit explicit uses of `Copyable` in the output module
for backwards compatability, we instead make it an error to use `Copyable`
as a type when not suppressed. While it's straightforward to omit it for
generic parameters, for uses of `Copyable` as an existential, we'd need to
emit `Any` in its place. We'd also need to find and omit or modify
explicit type constraints or same-type requirements involving Copyable.

All of that sounded rather gross, and not really needed right now since
any uses of `Copyable` explicitly is just not needed.

update test with Copyable being banned generally

kavon commented:
@swift-ci please test

kavon commented:
https://github.com/apple/swift-syntax/pull/1588
@swift-ci please test

kavon committed:
allow Copyable only in stdlib, with warning

kavon commented:
https://github.com/apple/swift-syntax/pull/1588
@swift-ci please test

kavon commented:
Just some minor test issues. I think this is ready for review.

tshortli commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2518,6 +2518,11 @@ void PrintAST::printInherited(const Decl *decl) {
+    // Printing a SuppressedTypeRepr will add the '~', so only print it
+    // if we will not be using the Repr's print method for this type.
+    if (!Options.PreferTypeRepr && inherited.isSuppressed)
+      Printer << "~";
```
We'll need to emit a feature guard when printing these suppressed constraints in swiftinterfaces in order to avoid confails with older compilers that don't understand the new operator.

kavon replied:
Good catch. Despite the only valid uses of them today being guarded by `$MoveOnly`, previous compilers that understood `$MoveOnly` did not also understand the `~` syntax.

kavon committed:
add `$SuppressedConformances` feature flag

This will help prevent condfails when older compilers try
to typecheck interfaces with `~Swift.Copyable` in them.

kavon commented:
https://github.com/apple/swift-syntax/pull/1588
@swift-ci please test

kavon commented:
@swift-ci please test

kavon commented:
https://github.com/apple/swift-syntax/pull/1588
@swift-ci please smoke test

kavon commented:
@swift-ci please smoke test

kavon committed:
clean-ups

kavon commented on lib/Sema/TypeCheckRequestFunctions.cpp:
```diff
@@ -82,6 +89,13 @@ Type InheritedTypeRequest::evaluate(
+    Evaluator &evaluator,
+    TypeOrExtensionDeclPointer decl,
+    unsigned index, TypeResolutionStage stage) const {
+  return resolveTypeLocWithin(decl, index, stage, getInheritedEntryAtIndex);
```
could revert this refactoring as it's not required anymore.

kavon commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -3385,6 +3390,12 @@ TypeResolver::resolveASTFunctionTypeParams(TupleTypeRepr *inputRepr,
           compileTimeConst = true;
           nestedRepr = specifierRepr->getBase();
           continue;
+        case TypeReprKind::Suppressed:
+          // Can't have a suppressed type as a function param.
+          diagnose(specifierRepr->getLoc(), diag::suppression_illegal_here);
```
ensure test coverage of this; forget if i added it

kavon commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -4314,6 +4345,39 @@ TypeResolver::resolveCompileTimeConstTypeRepr(CompileTimeConstTypeRepr *repr,
 }
 
+NeverNullType
+TypeResolver::resolveSuppressedTypeRepr(SuppressedTypeRepr *repr,
```
what happens if you tried
```
struct S: ~U {}
struct U: ~S {}
```

kavon commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -2902,24 +2902,27 @@ class DeclDeserializer {
-      // The low bit indicates "@unchecked".
-      bool isUnchecked = rawID & 0x01;
-      TypeID typeID = rawID >> 1;
+      // The first bit indicates "@unchecked", second bit indicates suppression.
+      bool isUnchecked = rawID & 0b01;
+      bool isSuppressed = rawID & 0b10;
+      TypeID typeID = rawID >> 2;
```
NOTE: this is why reading an old module would be impossible.

kavon commented on stdlib/public/core/CompilerProtocols.swift:
```diff
@@ -972,3 +972,19 @@ public protocol _ExpressibleByFileReferenceLiteral {
+/// For older compilers interacting with newer stdlibs, define the unavailable
+/// `_Copyable` marker protocol that they expect to find.
+@available(*, unavailable)
+typealias _Copyable = Copyable
```
must be a public typealias! I also wonder if it being a typealias even works. to play it safe maybe just duplicate the old one right here:
```
/// This marker protocol represents types that support copying.
/// This type is not yet available for use to express explicit
/// constraints on generics in your programs. It is currently
/// only used internally by the compiler.
@available(*, unavailable)
@_marker public protocol _Copyable {}
```

kavon reviewed (commented):
mostly NFC feedback but also noticed some issues.

kavon commented:
@swift-ci please test
